src/ctc_core.c

```diff
@@ -0,0 +1,462 @@
+#include <stdlib.h>
+#include <regex.h>
+#include "ctc_core.h"
+
+pthread_once_t ctc_api_once_init = PTHREAD_ONCE_INIT;
+
+CTC_HANDLE ctc_pool[MAX_CTC_HANDLE_COUNT];
+
+void ctc_api_init (void)
+{
+    int i, j;
+
+    for (i = 0; i < MAX_CTC_HANDLE_COUNT; i ++)
+    {
+        ctc_pool[i].ID = i;
+        ctc_pool[i].control_session.session_gid = -1;
+
+        for (j = 0; j < MAX_JOB_HANDLE_COUNT; j ++)
+        {
+            ctc_pool[i].job_pool[j].ID = j;
+            ctc_pool[i].job_pool[j].job_session.job_desc = -1;
+        }
+    }
+}
+
+int alloc_ctc_handle (CTC_HANDLE **ctc_handle_p)
+{
+    int i;
+
+    *ctc_handle_p = NULL;
+
+    for (i = 0; i < MAX_CTC_HANDLE_COUNT; i ++)
+    {
+        if (ctc_pool[i].control_session.session_gid == -1)
+        {
+            *ctc_handle_p = &ctc_pool[i];
+            
+            break;
+        }
+    }
+
+    if (IS_NULL (*ctc_handle_p))
+    {
+        goto error;
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int free_ctc_handle (CTC_HANDLE *ctc_handle)
+{
+    // 소켓 정리, 할당된 job 정리
+
+    ctc_handle->control_session.session_gid = -1;
+
+    return CTC_SUCCESS;
+}
+
+int alloc_job_handle (CTC_HANDLE *ctc_handle, JOB_HANDLE **job_handle_p)
+{
+    int i;
+
+    *job_handle_p = NULL;
+
+    for (i = 0; i < MAX_JOB_HANDLE_COUNT; i ++)
+    {
+        if (ctc_handle->job_pool[i].job_session.job_desc == -1)
+        {
+            *job_handle_p = &ctc_handle->job_pool[i];
+            
+            break;
+        }
+    }
+
+    if (IS_NULL (*job_handle_p))
+    {
+        goto error;
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int free_job_handle (JOB_HANDLE *job_handle)
+{
+    job_handle->job_session.job_desc = -1;
+
+    return CTC_SUCCESS;
+}
+
+int find_ctc_handle (int ctc_handle_id, CTC_HANDLE **ctc_handle_p)
+{
+    CTC_HANDLE *ctc_handle;
+
+    ctc_handle = &ctc_pool[ctc_handle_id];
+
+    if (ctc_handle->control_session.session_gid == -1)
+    {
+        goto error;
+    }
+
+    *ctc_handle_p = ctc_handle;
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int find_job_handle (CTC_HANDLE *ctc_handle, int job_handle_id, JOB_HANDLE **job_handle_p)
+{
+    JOB_HANDLE *job_handle;
+
+    job_handle = &ctc_handle->job_pool[job_handle_id];
+
+    if (job_handle->job_session.job_desc == -1)
+    {
+        goto error;
+    }
+
+    *job_handle_p = job_handle;
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int validate_url (char *url, CTC_HANDLE *ctc_handle)
```

Comment:
If this function intend doing both syntax check of 'url' and copy 'ip'/'port' to CTC_HANDLE,
I recommend change of the function name to the one which can express the behavior more suitably.

Reply:
We had already talked about this. I have two options and have not decided yet.
- change the function name correctly
- remove the "copy IP / Port" from the validate_url function